In [1]:
import csv
import xlrd
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt

# CPS Historical Migration/Geographic Mobility Tables
https://www.census.gov/data/tables/time-series/demo/geographic-mobility/historic.html
- Note: 
    - As of November 2018, these historical tables include corrections to errors that were discovered. 
    - Please replace any previously downloaded historical tables.
- Action:
    - load in tables
    - brief pre-exploration overview analysis (6 total)

In [2]:
# Table A-1. Annual Geographical Mobility Rates, By Type of Movement: 2018
annual_geographical_mobility_rates_type_movement = pd.read_excel( '../data/cps-historical-migration-geographic-mobility/tab-a-1.xls' )

# Table A-2. Annual Inmigration, Outmigration, Net Migration, and Movers from Abroad for Regions: 2018
annual_inmigration_outmigration_net_migration__movers_from_abroad_for_regions = pd.read_excel( '../data/cps-historical-migration-geographic-mobility/tab-a-2.xls' )

# Table A-3. Inmigration, Outmigration, and Net Migration by Metropolitan Status: 2018
inmigration_outmigration__net_migration_metropolitan_status = pd.read_excel( '../data/cps-historical-migration-geographic-mobility/tab-a-3.xls' )

# Table A-4. Geographical Mobility by Tenure: 2018
geographical_mobility_by_tenure = pd.read_excel( '../data/cps-historical-migration-geographic-mobility/tab-a-4.xls' )

# Table A-5. Reason for Move (Specific Categories): 2018
reason_for_move_specific_categories = pd.read_excel( '../data/cps-historical-migration-geographic-mobility/tab-a-5.xls' )

# Table A-6. Distance of Intercounty Move: 2018
distance_of_intercounty_move = pd.read_excel( '../data/cps-historical-migration-geographic-mobility/tab-a-6.xls' )

In [3]:
'''annual_geographical_mobility_rates_type_movement'''

print( annual_geographical_mobility_rates_type_movement.info() )
# annual_geographical_mobility_rates_type_movement.sample( 3 ) 

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 186 entries, 0 to 185
Data columns (total 10 columns):
Table with row headers in column A and column headers in rows 5 through 7.    154 non-null object
Unnamed: 1                                                                    137 non-null object
Unnamed: 2                                                                    137 non-null object
Unnamed: 3                                                                    137 non-null object
Unnamed: 4                                                                    138 non-null object
Unnamed: 5                                                                    137 non-null object
Unnamed: 6                                                                    138 non-null object
Unnamed: 7                                                                    137 non-null object
Unnamed: 8                                                                    137 non-null object
Unnamed: 9   

***annual_geographical_mobility_rates_type_movement***
- notes:
    - seems to cover back to 1947
    - needs to be cleaned , pd does not naturally detect formatting 
        - top and bottom
    - varying length of information
        - count (column(s))
            - 154 (0) , 138 (4, 6) , 137 (1, 2, 3, 5, 7, 8, 9)
        - information length is most likely <= 137

In [4]:
'''annual_inmigration_outmigration_net_migration__movers_from_abroad_for_regions'''

# print( annual_inmigration_outmigration_net_migration__movers_from_abroad_for_regions.info() )
annual_inmigration_outmigration_net_migration__movers_from_abroad_for_regions.sample( 3 )

Table with row headers in column A and column headers in row 5.  \
277                                .Movers from abroad                
111                                        .Inmigrants                
10                   .Net migration (including abroad)                

    Unnamed: 1 Unnamed: 2 Unnamed: 3 Unnamed: 4  
277        192        149        323        315  
111        362        679       1335        718  
10        -124        118       959*       212*

***annual_inmigration_outmigration_net_migration__movers_from_abroad_for_regions***
- notes:
    - seems to cover back to 1980
    - needs to be cleaned , pd does not naturally detect formatting 
        - top and bottom

In [5]:
'''inmigration_outmigration__net_migration_metropolitan_status'''

# print( inmigration_outmigration__net_migration_metropolitan_status.info() )
inmigration_outmigration__net_migration_metropolitan_status.sample( 3 )

Table with row headers in column A and column headers in row 5. Unnamed: 1  \
59                             .Nonmetropolitan areas                     977   
23                                        2014-2015/1                     NaN   
27                                 ..Principal cities                    3455   

   Unnamed: 2 Unnamed: 3 Unnamed: 4 Unnamed: 5  
59       1151      -175*         63       -111  
23        NaN        NaN        NaN        NaN  
27       5198    -1,744*        515    -1,229*

***inmigration_outmigration__net_migration_metropolitan_status***
- notes:
    - seems to cover back to 1985
    - needs to be cleaned , pd does not naturally detect formatting 
        - top and bottom

In [6]:
'''geographical_mobility_by_tenure'''

# print( geographical_mobility_by_tenure.info() )
geographical_mobility_by_tenure.sample( 3 )

Table with row headers in column A and column headers in rows 5 through 7.  Numbers start in row 10, percentages start in row 184, and frequency distributions start in row 358.  \
341                          ..In owner-occupied units                                                                                                                                 
56                          ..In renter-occupied units                                                                                                                                 
237                                          2008-2009                                                                                                                                 

    Unnamed: 1 Unnamed: 2 Unnamed: 3 Unnamed: 4 Unnamed: 5 Unnamed: 6  \
341        100       90.8        9.2          9        5.2        3.8   
56       93899      67177      26722      26040      18951       7089   
237        NaN        NaN        NaN        NaN        NaN        NaN   

    Unnamed: 7 Unnamed: 8 Unnamed: 9  
341          2        1.8        0.2  
56        4183       2906        682  
237        NaN        NaN        NaN

***geographical_mobility_by_tenure***
- notes:
    - seems to cover back to 1987
    - needs to be cleaned , pd does not naturally detect formatting 
        - top and bottom

In [7]:
'''reason_for_move_specific_categories'''

# print( reason_for_move_specific_categories.info() )
reason_for_move_specific_categories.sample( 3 )

Table with row headers in columns A and K, and column headers in rows 5 and 6. Numbers start in row 9, percentages start in row 36.  \
11                  2013-2014 (98,000 address file)/2                                                                                    
0   Table A-5. Reason for Move (Specific Categorie...                                                                                    
24                                          2004-2005                                                                                    

   Unnamed: 1 Unnamed: 2 Unnamed: 3 Unnamed: 4 Unnamed: 5 Unnamed: 6  \
11      35681       1742       3966       4797       3471        745   
0         NaN        NaN        NaN        NaN        NaN        NaN   
24      39888       2841       3108       4882       4157        758   

   Unnamed: 7 Unnamed: 8 Unnamed: 9  ... Unnamed: 12 Unnamed: 13 Unnamed: 14  \
11       2211        249        694  ...        5629        1054        3356   
0         NaN        NaN        NaN  ...         NaN         NaN         NaN   
24       1366        216        540  ...        7091        1609        2648   

   Unnamed: 15 Unnamed: 16 Unnamed: 17 Unnamed: 18 Unnamed: 19 Unnamed: 20  \
11         477        4578         177          31         129          10   
0          NaN         NaN         NaN         NaN         NaN         NaN   
24           -        3759        1289         234         633           -   

   Unnamed: 21  
11         359  
0          NaN  
24        1052  

[3 rows x 22 columns]

***reason_for_move_specific_categories***
- notes:
    - seems to cover back to 1998
    - needs to be cleaned , pd does not naturally detect formatting 
        - top and bottom

In [8]:
# distance_of_intercounty_move
# print( distance_of_intercounty_move.info() )
distance_of_intercounty_move.sample( 3 )

Table with row headers in column A and column headers in row 5. Numbers start in row 8, percentages start in row 32.  \
22                                          2005-2006                                                                     
1                             (Numbers in thousands.)                                                                     
33                                                NaN                                                                     

   Unnamed: 1 Unnamed: 2 Unnamed: 3 Unnamed: 4 Unnamed: 5  
22      13690       5595       2910       2092       3092  
1         NaN        NaN        NaN        NaN        NaN  
33        NaN        NaN        NaN        NaN        NaN

***distance_of_intercounty_move***
- notes:
    - seems to cover back to 2002
    - needs to be cleaned , pd does not naturally detect formatting 
        - top and bottom

## Aggregate Initial Notes
- ***Dates***
    - first recording year varies by up to 55 years 

| first recording year | dataset |
|- - - -|- - - -|
| 1947 | annual_geographical_mobility_rates_type_movement |
|- - - -|- - - -|
| 1980 | annual_inmigration_outmigration_net_migration__movers_from_abroad_for_regions |
|- - - -|- - - -|
| 1985 | inmigration_outmigration__net_migration_metropolitan_status |
|- - - -|- - - -|
| 1987 | inmigration_outmigration__net_migration_metropolitan_status |
|- - - -|- - - -|
| 1998 | reason_for_move_specific_categories |
|- - - -|- - - -|
| 2002 | distance_of_intercounty_move |

- ***format***
    - first rows of each file seem to serve as information , actual tables start near end of .head()
    - ending rows of each file seem to serve as annotation and disclaimer
- ***actions***:
    - walk though one table at a time
    - let's start with dropping NaN clustered rows

In [9]:
# copy dataframe , generalize name
df = annual_geographical_mobility_rates_type_movement.copy()
# drop columns without any non-NaN values
df = df.dropna(thresh= ( 1 ) )
# return the dataframe to see results
# df
df.sample(7)

Table with row headers in column A and column headers in rows 5 through 7.  \
174  1/ The 2014 CPS ASEC included redesigned quest...                           
149                                          1966-1967                           
129                                          1990-1991                           
81                                           1953-1954                           
1                              (Numbers in thousands.)                           
3                                      Mobility period                           
122                                          1995-1996                           

                     Unnamed: 1                   Unnamed: 2    Unnamed: 3  \
174                         NaN                          NaN           NaN   
149                         100                           81            19   
129                         100                           83            17   
81                       155679                       125654         30025   
1                           NaN                          NaN           NaN   
3    Total, 1 year old and over  Same residence (non-movers)  Total movers   
122                         100                         83.7          16.3   

                                   Unnamed: 4 Unnamed: 5 Unnamed: 6  \
174                                       NaN        NaN        NaN   
149                                      18.3       11.6        6.7   
129                                      16.4       10.3        6.1   
81                                      29027      19046       9981   
1                                         NaN        NaN        NaN   
3    Different residence in the United States        NaN        NaN   
122                                      15.8       10.3        5.6   

    Unnamed: 7 Unnamed: 8          Unnamed: 9  
174        NaN        NaN                 NaN  
149        3.3        3.4                 0.7  
129        3.2        2.9                 0.6  
81        4947       5034                 998  
1          NaN        NaN                 NaN  
3          NaN        NaN  Movers from abroad  
122        3.1        2.5                 0.5

- ***notes:***
    - 156 rows remain
    - some column names in row 3 , some column names in row 4 , some column names in row 5
        row 3 = 6 column names
            columns = 0 , 1 , 2 , 3 , 4 , 9
        row 4 = 2 column names
            columns = 5 , 6
        row 5 = 2 column names
            columns = 7 , 8
    - row 4 column 4 and row 5 column 6 contain 'Total'
        not obvious what this is
        but it is not the column name in either case
        in both cases , the column name is one row higher in the same column
    - rows 173 - 185 contain only 1 NaN value
- ***actions:***
    - drop all rows without 2 or more non-NaN values
    - merge rows 3 , 4 , and 5 
        taking out 'Total'

In [10]:
# rename dataframe
df = annual_geographical_mobility_rates_type_movement.copy()

# drop columns with less than 2 non-NaN values
df = df.dropna( thresh= ( 2 ) )

# collect column names as pandas series
col_names_0to4_9 = df.iloc[ 0 ]
col_names_5to6 = df.iloc[ 1 ]
col_names7to8 = df.iloc[ 2 ]

# initialize list for column names
column_names = []

# fill list with first row of column names , no exceptions
for column_name in col_names_0to4_9:    
    column_names.append( column_name )

# set index count to -1 so column 0 == column 0
_ = -1

# replace items 5 and 6 in column_names with column names 5 and 6
for column_name in col_names_5to6:
    _ += 1
    # column_name is string (nan instances = float) and name is not 'Total'   
    if isinstance( column_name , str ) and column_name != 'Total':
        # column_name is aligned to replace a nan         
        if str( column_names[ _ ] ) == 'nan':
            # replace item in column_names with column_name             
            column_names[ _ ] = column_name

# set index count to -1 so column 0 == column 0
_ = -1

# replace items 7 and 8 in column_names with column names 7 and 8
for column_name in col_names7to8:
    _ += 1
    # column_name is string (nan instances = float) and name is not 'Total'   
    if isinstance( column_name , str ) and column_name != 'Total':
        # column_name is aligned to replace a nan         
        if str( column_names[ _ ] ) == 'nan':
            # replace item in column_names with column_name             
            column_names[ _ ] = column_name

# (re)set column names
df.columns = column_names

# drop rows used to extract column names
df = df[3:]

# reset dataframe index
df = df.reset_index( drop=True ) 

# let's see
df.head(10)

Mobility period Total, 1 year old and over  \
0                          2017-2018                     319319   
1                          2016-2017                     316550   
2                          2015-2016                     314992   
3                          2014-2015                     312295   
4  2013-2014 (98,000 address file)/1                     309601   
5                          2012-2013                     307243   
6                          2011-2012                     304924   
7        2010-2011 (2010 controls)/2                     302640   
8        2010-2011 (2000 controls)/3                     302005   
9        2009-2010 (2010 controls)/2                     300419   

  Same residence (non-movers) Total movers  \
0                      286967        32352   
1                      281647        34902   
2                      279854        35138   
3                      275971        36324   
4                      273920        35681   
5                      271325        35918   
6                      268436        36488   
7                      267602        35038   
8                      266930        35075   
9                      262975        37445   

  Different residence in the United States Same county Different county  \
0                                    31186       19852            11334   
1                                    33647       21614            12033   
2                                    33857       21588            12269   
3                                    34651       22905            11746   
4                                    34548       23436            11112   
5                                    34882       23150            11731   
6                                    35334       23493            11842   
7                                    33953       23330            10623   
8                                    34016       23325            10691   
9                                    36459       25910            10549   

  Same state Different state Movers from abroad  
0       6496            4838               1165  
1       6668            5366               1255  
2       7501            4768               1281  
3       6653            5093               1673  
4       6446            4666               1133  
5       6961            4770               1036  
6       6782            5059               1154  
7       5868            4756               1084  
8       5912            4779               1058  
9       6227            4323                985

- ***as of now***:
    - removed rows with NaN values
    - extracted, sorted, and set column names
- ***next (action)***:
    - we need to clean up the 'Mobility period' column
        - samples from current:
            - 2017-2018
            - 2013-2014 (98,000 address file)/1
            - 2011-2012
            - 2010-2011 (2010 controls)/2
            - 2000-2001 (SCHIP, 2000 controls)/5
            - 1970-1971/8
        - what we want:
            - 2018 << 2017-2018
            - 2014 << 2013-2014 (98,000 address file)/1
                requires further investigation
            - 2012 << 2011-2012
            - 2011 << 2010-2011 (2010 controls)/2
                requires further investigation
            - 2001 << 2000-2001 (SCHIP, 2000 controls)/5
                requires further investigation
            - 1971 << 1970-1971/8

In [11]:
# examine data in 'Mobility period'
# _ = 0
# for instance in df[ 'Mobility period' ]:
#     # print( _ , instance )
#     _ += 1
    
# selected outputs 
'''
# non-uniform , non-repeated
    3 2014-2015
    4 2013-2014 (98,000 address file)/1
    5 2012-2013
    
# repeats w/ controls , no uniform representation
    6 2011-2012
    7 2010-2011 (2010 controls)/2
    8 2010-2011 (2000 controls)/3
    9 2009-2010 (2010 controls)/2
    10 2009-2010 (2000 controls)/3
    11 2008-2009

    18 2001-2002
    19 2000-2001 (2000 controls)/3
    20 2000-2001 (SCHIP, 2000 controls)/5
    21 2000-2001 (1990 controls)/6
    22 1999-2000
    
    28 1993-1994
    29 1992-1993 (1990 controls)/6
    30 1992-1993 (1980 controls)/7
    31 1991-1992
    
# lack of info from 70s 
    41 1981-1982
    42 1980-1981
    43 1975-1976/8
    44 1970-1971/8
    45 1969-1970
    46 1968-1969
    47 1967-1968
    
# in chronological order until row 68 (then repeats)
    66 1948-1949
    67 1947-1948
    68 2017-2018
    69 2016-2017
'''

# make sure rows 0-67 are the same as 68-135
# new copy
copy_mobility_period = df[ 'Mobility period' ].copy()

# index first half
for _ in range( len( copy_mobility_period[ 0:68 ] ) ):
    # print each row in first half with corresponding row in 2nd half
    print( copy_mobility_period[ _ ] + '  |  ' + copy_mobility_period[ 68 + _ ])

2017-2018  |  2017-2018
2016-2017  |  2016-2017
2015-2016  |  2015-2016
2014-2015  |  2014-2015
2013-2014 (98,000 address file)/1  |  2013-2014 (98,000 address file)/1
2012-2013  |  2012-2013
2011-2012  |  2011-2012
2010-2011 (2010 controls)/2  |  2010-2011 (2010 controls)/2
2010-2011 (2000 controls)/3  |  2010-2011 (2000 controls)/3
2009-2010 (2010 controls)/2  |  2009-2010 (2010 controls)/2
2009-2010 (2000 controls)/3  |  2009-2010 (2000 controls)/3
2008-2009  |  2008-2009
2007-2008  |  2007-2008
2006-2007  |  2006-2007
2005-2006  |  2005-2006
2004-2005/4  |  2004-2005/4
2003-2004  |  2003-2004
2002-2003  |  2002-2003
2001-2002  |  2001-2002
2000-2001 (2000 controls)/3  |  2000-2001 (2000 controls)/3
2000-2001 (SCHIP, 2000 controls)/5  |  2000-2001 (SCHIP, 2000 controls)/5
2000-2001 (1990 controls)/6  |  2000-2001 (1990 controls)/6
1999-2000  |  1999-2000
1998-1999  |  1998-1999
1997-1998  |  1997-1998
1996-1997  |  1996-1997
1995-1996  |  1995-1996
1994-1995  |  1994-1995
1993-1994 

- ***notes***:
    - table repeats time periods
- ***action***:
    - save this for later, focus on conversion of year ranges to years

In [12]:
# determine type of data in 'Mobility period' 
def test_type( dataframe , test_type ):
    '''
    input) dataframe w/ specified column 
        - e.g. df['specified column']
    input) type wanting to test
        - e.g. str 
    
    output) count of instances which are not type == test_type
    '''
    non_test_type = 0
    for _ in dataframe:
        if isinstance( _ , test_type ) == False:
            non_test_type += 1
    return non_test_type


test_000 = test_type( dataframe=df[ 'Mobility period' ] , test_type=str )
test_000

0

- ***as test_000 returns 0***:
    - the type of all datapoints in 'Mobility period' is str (string)
    - we will not need to change any points to have uniformity   
- ***next up (action)***:
    - clear low-hanging fruit
    - change the simple ranges to int for set year
        - e.g. change "2017-2018" to "2018"

In [13]:
# identify changes which can be implimented without further investigation
mobility_period = df[ 'Mobility period' ]
simple_year_ranges_count_ = 0
confirmed_simple_year_ranges_count_ = 0

# count how many simple-fix years we have
for datapoint in mobility_period:
    # len('2012-2013') == 9 , check datapoint is formatted as expected     
    if len( datapoint ) == 9 and len( datapoint.split( '-' ) ) == 2:
        # pretty sure we know what's going on but will investigate further         
        simple_year_ranges_count_ += 1
        # first 4 characters of string , can be converted to int
        first_year = int( datapoint[ :4 ] )
        # last 4 characters of string , can be converted to int
        second_year = int( datapoint[ -4: ] )
        # first_year and last_year are post-1940 (because expected), and year range is 1 year 
        if first_year and second_year > 1940 and second_year - first_year == 1:
            # the datapoint is within our risk threshold
            confirmed_simple_year_ranges_count_ += 1

# if everything we thought was what we think it is 
if simple_year_ranges_count_ == confirmed_simple_year_ranges_count_:
    print(f'total: {confirmed_simple_year_ranges_count_}\nratio: {confirmed_simple_year_ranges_count_/len(mobility_period)}')

total: 110
ratio: 0.8088235294117647


- ***notes***:
    - know count of lowest fruit to be 110 ( > 80.88% , not bad)
    - just confirmed status of 110 points to be (at least) similar enough to extract years
    - knowing simple_year_ranges_count_ allows the conformation of outcome via count compairson 
- ***action***:
    - convert those 110 year ranges to years represented by the 2nd year in each range

In [14]:
# set empty
less_low_mobility_periods = []
# so row 0 == row 0
_ = -1

# little bit of ctrl + c , little bit of ctrl + v
for datapoint in mobility_period:
    _ += 1
    # len('2012-2013') == 9 , check datapoint is formatted as expected     
    if len( datapoint ) == 9 and len( datapoint.split( '-' ) ) == 2:
        # replace datapoint w/ int of last 4 characters of string 
        mobility_period[ _ ] = int( datapoint[ -4: ] )
    else:
        # keep what's already there
        less_low_mobility_periods.append( datapoint )

mobility_period.sample(7)

81                            2007
104                           1987
49                            1966
63                            1952
76     2010-2011 (2000 controls)/3
132                           1951
60                            1955
Name: Mobility period, dtype: object

- ***notes***:
    - we have converted the low hanging fruit
- ***actions***:
    - now on to exploring the other 26 rows

In [15]:
# print them out , let's see what we can see
_ = 0
for row in df['Mobility period']:
    _ += 1    
    if not isinstance( row , int ):
        # _ check for consistent spacing
        if _ < 10:
            print( str(_) + '  |  ' + row )
        elif _ < 100:
            print( str(_) + ' |  ' + row )
        else:
            print( str(_) + '|  ' + row )

5  |  2013-2014 (98,000 address file)/1
8  |  2010-2011 (2010 controls)/2
9  |  2010-2011 (2000 controls)/3
10 |  2009-2010 (2010 controls)/2
11 |  2009-2010 (2000 controls)/3
16 |  2004-2005/4
20 |  2000-2001 (2000 controls)/3
21 |  2000-2001 (SCHIP, 2000 controls)/5
22 |  2000-2001 (1990 controls)/6
30 |  1992-1993 (1990 controls)/6
31 |  1992-1993 (1980 controls)/7
44 |  1975-1976/8
45 |  1970-1971/8
73 |  2013-2014 (98,000 address file)/1
76 |  2010-2011 (2010 controls)/2
77 |  2010-2011 (2000 controls)/3
78 |  2009-2010 (2010 controls)/2
79 |  2009-2010 (2000 controls)/3
84 |  2004-2005/4
88 |  2000-2001 (2000 controls)/3
89 |  2000-2001 (SCHIP, 2000 controls)/5
90 |  2000-2001 (1990 controls)/6
98 |  1992-1993 (1990 controls)/6
99 |  1992-1993 (1980 controls)/7
112|  1975-1976/8
113|  1970-1971/8


- ***notes***
    - 6 rows formatted almost like majority (new lowest)
        - all have either /4 or /8 at the end of otherwise standard string
            - e.g. standard 
                - 2004-2005
            - observation rows 
                - 16 |  2004-2005/4
                - 44 |  1975-1976/8
                - 45 |  1970-1971/8
                - 84 |  2004-2005/4
                - 112|  1975-1976/8
                - 113|  1970-1971/8
        - remarks before reading documentation
            - identical instances
                - rows 16 and 84
                - rows 44 and 112
                - rows 45 and 113
            - rows are not ordered by year
                - unless years reoccour/loop through table
                - or otherwise
    - 16 rows are in the next new lowest format
        - all contain ({year} controls)/{num} with year and num in consistent pairs
            - (year / num) pairs = # occurrences
                - (2010 / 2) = 4
                - (2000 / 3) = 6
                - (1990 / 6) = 4
                - (1980 / 7) = 2
        - they occour in pattern twice:
            - 4 uninterrupted
            - 1
            - 3 uninterrupted
        - sample rows
            - 8  |  2010-2011 (2010 controls)/2
            - 9  |  2010-2011 (2000 controls)/3
            - 10 |  2009-2010 (2010 controls)/2
            - 11 |  2009-2010 (2000 controls)/3
        - unknown analysis
            - seems to be game settings for different censuses 
            - difficult to connect the /2,/3 and /6,/7 relationship
                - if was dependent on decade order , possibles would be straight ordered 
                    - e.g. /2, /3, /4, /5 or /1, /3, /5, /7
                    
    - final 4 rows are longer and are 2 duplicate pairs
        - default format + ending either (98,000 address file)/1 or (SCHIP, 2000 controls)/5
            - longest rows
                - 5  |  2013-2014 (98,000 address file)/1
                - 21 |  2000-2001 (SCHIP, 2000 controls)/5
                - 73 |  2013-2014 (98,000 address file)/1
                - 89 |  2000-2001 (SCHIP, 2000 controls)/5
        - pre document remarks
            - first notice of /1 
            - return use of /5 
                - do other /n have additional possibile (paramaters)
    - blind analysis
        - year range
                - min: 1970-1971 (1971 in complete df)
                - max: 2013-2014 (2014 in complete df)
            - less range than full representation of column
                - min: 1948 (1947-1948 in origional df)
                - max: 2018 (2017-2018 in origional df)
            - why?
                - recent changes to census
                    - collection
                    - questions
                    - coverage
                - recent increase in census related surveys
                    - housing surveys
                    - business surveys
                        - we will be exploring these after fitting gen_000 model
                        - these are also how this data is presented for non-decade years
        - lot of possible meanings for non-conforming objects
                - politics..
                - changing standards
                - technological evolution
                - etc...
                    - storage , collection , methods , etc... 
            - full-court lobs
                - something to do with immigrants
                - evolution of process 
                - expansion of non-physical opportunities 
        - there's got to be a better way to extract dataframes from excel files
            - likely convertable, but annotations likely direct either universe to same result 
- ***actions***
    - go find out what these non-standard occurrences mean

In [16]:
# make another copy of origional dataframe (this is why we copy)
info_df = annual_geographical_mobility_rates_type_movement.copy()

# function to make NaN dataframe
row_nans = lambda df: df[ df.isnull().any( axis=1 ) ]

# identify the rows with information (basically, keep the NaNs this time)
nan_df = row_nans( info_df )

# peak some rows with NaN values 
nan_df.head()

Table with row headers in column A and column headers in rows 5 through 7.  \
0  Table A-1. Annual Geographical Mobility Rates,...                           
1                            (Numbers in thousands.)                           
2                                                NaN                           
3                                    Mobility period                           
4                                                NaN                           

                   Unnamed: 1                   Unnamed: 2    Unnamed: 3  \
0                         NaN                          NaN           NaN   
1                         NaN                          NaN           NaN   
2                         NaN                          NaN           NaN   
3  Total, 1 year old and over  Same residence (non-movers)  Total movers   
4                         NaN                          NaN           NaN   

                                 Unnamed: 4   Unnamed: 5        Unnamed: 6  \
0                                       NaN          NaN               NaN   
1                                       NaN          NaN               NaN   
2                                       NaN          NaN               NaN   
3  Different residence in the United States          NaN               NaN   
4                                     Total  Same county  Different county   

  Unnamed: 7 Unnamed: 8          Unnamed: 9  
0        NaN        NaN                 NaN  
1        NaN        NaN                 NaN  
2        NaN        NaN                 NaN  
3        NaN        NaN  Movers from abroad  
4        NaN        NaN                 NaN

In [17]:
# noted column 0 name 
column_names = []

# make it slightly educational
for _ in nan_df[ :0 ]:
    # each _ is a string
    if isinstance( _ , str ):
        # add the string to column_names
        column_names.append( _ )
        
# this is the education
print( f'len(nan_df[:0])=={ len( nan_df[ :0 ] ) }\ntype(nan_df[:0])=={ type( nan_df[ :0 ] ) }\n'
      # The economy, stupid. 
       f'len(column_names)=={ len( column_names ) }\ntype(column_names={ type( column_names ) }' )

len(nan_df[:0])==0
type(nan_df[:0])==<class 'pandas.core.frame.DataFrame'>
len(column_names)==10
type(column_names=<class 'list'>


In [18]:
# we can find row headers in column A
if column_names[0] != 'Table with row headers in column A and column headers in rows 5 through 7.':
    raise AssertionError('')

# function to return non-NaN rows 
non_nans = lambda df: df.dropna()

# identify non-NaN rows from column A (the first column)
column_a = non_nans( nan_df[ 'Table with row headers in column A and column headers in rows 5 through 7.' ] )

# let's see
column_a

0      Table A-1. Annual Geographical Mobility Rates,...
1                                (Numbers in thousands.)
3                                        Mobility period
6                                                 NUMBER
90                                               PERCENT
173                                           Footnotes:
174    1/ The 2014 CPS ASEC included redesigned quest...
175    2/ Population controls consistent with 2010 Ce...
176    3/ Population controls consistent with 2000 Ce...
177    4/ Caution should be used when comparing numbe...
178    5/ SCHIP sample: poulation controls consistent...
179    6/ Population controls consistent with 1990 Ce...
180    7/ Population controls consistent with 1980 Ce...
181    8/ The 1-year geographic mobility question was...
182    For more information on the transition from Ce...
183    Adjustments to Household Survey Population Est...
184       Population Estimates in January 2012 webpage\n
185    Source:  U.S. Census Bur

In [19]:
# identify information of interest
indicators_of_interest = [ '1/' , '2/' , '3/' , '4/' , '5/' , '6/' , '7/' , '8/' ]

# make copy
copy_column_a = column_a.copy()

# return rows from copy_column_a which contain an indicator from indicators_of_interest
rows_of_interest = [ row for indicator in indicators_of_interest for row in copy_column_a if indicator in row ]

# examine the lowest (remember: /4 and /8)
for row in rows_of_interest:
    if '4/' in row or '8/' in row:
        print( row + '\n' )

4/ Caution should be used when comparing numbers/rates of movers within the same county and from a different county between the 1999-2000 to 2004-2005 period with other periods. A change in processing resulted in a shift in the distribution of movers for these years. For additional information, see the "Impact of Processing on CPS Interstate Migration Rates" note < http://www.census.gov/population/www/socdemo/CPSnote.pdf >.   

8/ The 1-year geographic mobility question was not asked between 1972 through 1975 and 1977 to 1980. The first half of the decade (1971-1975) asked about migration since 1970 and the second half (1976-1980) since 1975. Therefore, only 1970-1971 and 1975-1976 ask a comparable 1-year question.    



- ***notes***:
    - /4 is fairly easy to account for
        - basically they flipped the method of recording 
            - https://www.census.gov/population/www/socdemo/CPSnote.pdf 
            - from
                - group > geography
            - to 
                - geography > group
        - "Users of interstate migration data should be cautious of comparing rates from the 1999-2000 to 2004-5 period with other periods."
    - /8 is dissapointing due to prelim scanning of San Francisco 
        - will need to be remembered if one year are factored
        - also indicates 5-year question could be used 
            - would need to ensure other 5-year periods are/get sorted 
            
- ***actions***:
    - compare with neighbors to determine optimal implementation 
        - hedge repeats

In [20]:
# reset
sims = df['Mobility period']
_ = -1

for row in sims:
    # update row
    _ += 1
    # id indicators
    if '/4'  in str(row) or '/8' in str(row): 
        # ensure next is logical
        if _ < 2:
            raise Exception( 'f' )
        # print 2 before through 2 after
        print ( '\n\n\n\n' , sims[ _-2 : _+3 ] )





 13           2007
14           2006
15    2004-2005/4
16           2004
17           2003
Name: Mobility period, dtype: object




 41           1982
42           1981
43    1975-1976/8
44    1970-1971/8
45           1970
Name: Mobility period, dtype: object




 42           1981
43    1975-1976/8
44    1970-1971/8
45           1970
46           1969
Name: Mobility period, dtype: object




 81           2007
82           2006
83    2004-2005/4
84           2004
85           2003
Name: Mobility period, dtype: object




 109           1982
110           1981
111    1975-1976/8
112    1970-1971/8
113           1970
Name: Mobility period, dtype: object




 110           1981
111    1975-1976/8
112    1970-1971/8
113           1970
114           1969
Name: Mobility period, dtype: object


- ***notes***:
    - no /4 and /8 instances have duplicate year range  
    - we are reliant on both /4 and /8 instances (all)
    
- ***actions***:
    - continue examining other non-standard rows
    - add /4 and /8 to list for default conversion 
        - default conversion = note in new column & convert as if standard

In [21]:
# make copy of rows of interest
copy_rows_of_interest = rows_of_interest.copy()
# used: /4, /8 ; unused: 1/, /2, /3, /5, /6, /7

# examine the next lowest (remember: /2 and /3 ; largest set, size=16)
for row in copy_rows_of_interest:
    if '2/' in row or '3/' in row:
        print( row + '\n' )

2/ Population controls consistent with 2010 Census.

3/ Population controls consistent with 2000 Census.



- ***notes***:
    - requires conformation of duplicates but ideally move forward with 2010 controls 
        - values most recent measures
        - return for 2000 controls for comps if time
    - https://www.census.gov/content/dam/Census/programs-surveys/acs/guidance/2010_Change_Population_Controls.pdf 
- ***actions***:
    - confirm duplicates 

In [22]:
# reset
sims_001 = df['Mobility period']

# for each instance in sims_001
for _ in range( len( sims_001 ) ):
    # if instance is string
    if isinstance( sims_001[ _ ] , str ):
        # and first 11 characters in that string are same as first 11 characters in str(next row)
            # e.g. 2010-2011 ( == 2010-2011 (
        if sims_001[ _ ][ :11 ] == str( sims_001[ _+1 ])[ :11 ]:
            # output the first row, then the second
            print( sims[ _ ], '\n', sims[ _+1 ], '\n' )

2010-2011 (2010 controls)/2 
 2010-2011 (2000 controls)/3 

2009-2010 (2010 controls)/2 
 2009-2010 (2000 controls)/3 

2000-2001 (2000 controls)/3 
 2000-2001 (SCHIP, 2000 controls)/5 

2000-2001 (SCHIP, 2000 controls)/5 
 2000-2001 (1990 controls)/6 

1992-1993 (1990 controls)/6 
 1992-1993 (1980 controls)/7 

2010-2011 (2010 controls)/2 
 2010-2011 (2000 controls)/3 

2009-2010 (2010 controls)/2 
 2009-2010 (2000 controls)/3 

2000-2001 (2000 controls)/3 
 2000-2001 (SCHIP, 2000 controls)/5 

2000-2001 (SCHIP, 2000 controls)/5 
 2000-2001 (1990 controls)/6 

1992-1993 (1990 controls)/6 
 1992-1993 (1980 controls)/7 



- ***notes***:
    - duplicate confirmed:
        - 2010-1011
        - 2009-2010
        - 1992-1993        
        - each duplicate repeated in line with table repeat...
    - duplicate *NOT* confirmed:
        - /2 or /3 dependent 
            - 2000-2001
                - 2000-2001 (2000 controls)/3 
                - 2000-2001 (SCHIP, 2000 controls)/5 
            - duplicate repeated in line with table repeat...
        - *NOT* /2 or /3 dependent
            - 2000-2001
                - 2000-2001 (SCHIP, 2000 controls)/5 
                - 2000-2001 (1990 controls)/6 
            - 1992-1993
                - 1992-1993 (1990 controls)/6 
                - 1992-1993 (1980 controls)/7 
            - each duplicate repeated in line with table repeat...
    - some of these contain /5
        - *but NONE* contain /1
- ***actions***:
    - re-check for duplicate year ranges between other /# 
        - always good practice
        - could lead to reconsideration of one(+) /#

In [23]:
## reset
sims_002 = df[ 'Mobility period' ]
_ = 0
# to remember what we've seen
storage = []

# let's see what we can see
for row in sims_002:
    # keep track of the row
    _ += 1 

    # if instance (row) is not int
    if not isinstance( row , int ):
        
        # add 2nd year in year range to storage
        storage.append( row[ 5:9 ] )
        
        # how many times have we seen that year
        s_c_ = str( storage.count( row[ 5:9 ] ) )
        
        # we only care if the year has been seen at least twice (inclusive) as all rows are seen twice
        if int( s_c_ ) >= 2:
            # print row number, 2nd year in year range, times 2nd year has been seen, entire row
            # account for row number when formatting
            if _ < 10:
                print( str( _ ) + '  |  ' + row[ 5:9 ] + '  |  ' + s_c_ + '  |  ' + row )
            elif _ < 100:
                print( str( _ ) + ' |  ' + row[ 5:9 ] + '  |  ' + s_c_ + '  |  ' + row )
            else:
                print( str( _ ) + '|  ' + row[ 5:9 ] + '  |  ' + s_c_ + '  |  ' + row )

9  |  2011  |  2  |  2010-2011 (2000 controls)/3
11 |  2010  |  2  |  2009-2010 (2000 controls)/3
21 |  2001  |  2  |  2000-2001 (SCHIP, 2000 controls)/5
22 |  2001  |  3  |  2000-2001 (1990 controls)/6
31 |  1993  |  2  |  1992-1993 (1980 controls)/7
73 |  2014  |  2  |  2013-2014 (98,000 address file)/1
76 |  2011  |  3  |  2010-2011 (2010 controls)/2
77 |  2011  |  4  |  2010-2011 (2000 controls)/3
78 |  2010  |  3  |  2009-2010 (2010 controls)/2
79 |  2010  |  4  |  2009-2010 (2000 controls)/3
84 |  2005  |  2  |  2004-2005/4
88 |  2001  |  4  |  2000-2001 (2000 controls)/3
89 |  2001  |  5  |  2000-2001 (SCHIP, 2000 controls)/5
90 |  2001  |  6  |  2000-2001 (1990 controls)/6
98 |  1993  |  3  |  1992-1993 (1990 controls)/6
99 |  1993  |  4  |  1992-1993 (1980 controls)/7
112|  1976  |  2  |  1975-1976/8
113|  1971  |  2  |  1970-1971/8


- ***notes***:
    - issue
        - no omnipresent control
            - pre-2000 have 1990 or 1980 controls
            - post-2000 have 2000 or 2010 controls 
            - 2000 has 2000, 1990 and SCHIP 2000 controls
    - solutions
        - compare everything to 2000 
        - identify differences between 80, 90, 00, SCHIP 00, and 10
    - 2013-2014 is standalone, no options
- ***actions***:
    - restate /2 and /3 with /5, /6 and /7
    - check for differences , document
    - determine which filters to move forward with

In [24]:
# make copy of rows of interest
copy2_rows_of_interest = rows_of_interest.copy()
# used: /2, /3, /4, /8 ; unused: 1/, /5, /6, /7

re_state_with = ['2/' , '3/' , '5/' , '6/', '7/']
# examine the next lowest (remember: /2 and /3 ; largest set, size=16)
for row in copy2_rows_of_interest:
    for filter in re_state_with:
        if filter in row:
            print( row + '\n' )

2/ Population controls consistent with 2010 Census.

3/ Population controls consistent with 2000 Census.

5/ SCHIP sample: poulation controls consistent with 2000 Census.

6/ Population controls consistent with 1990 Census.

7/ Population controls consistent with 1980 Census.



In [25]:
# copy current dataframe (annual_geographical_mobility_rates_type_movement)
df_copy = df.copy()

# list of rows without converted years (reminder: origional 2010-2011 converted 2011)
non_converted_years = [row for row in df_copy[ 'Mobility period' ] if isinstance( row , str ) ]

# filter for rows which are not yet converted in the 'Mobility period' column
n_c_y_df = df_copy.loc[ df_copy['Mobility period'].isin( non_converted_years ) ]

# visualize (%) differences between rows 
i_mobility_change = n_c_y_df.set_index('Mobility period').pct_change()
# note: 0.0101 = 1.01% ; first row will be NaN ; compare 2nd same year to 1st (e.g. 2011 vs 2011 not 2011 vs 2013)
i_mobility_change

Total, 1 year old and over  \
Mobility period                                                  
2013-2014 (98,000 address file)/1                          NaN   
2010-2011 (2010 controls)/2                          -0.022484   
2010-2011 (2000 controls)/3                          -0.002098   
2009-2010 (2010 controls)/2                          -0.005252   
2009-2010 (2000 controls)/3                          -0.001148   
2004-2005/4                                          -0.043076   
2000-2001 (2000 controls)/3                          -0.040178   
2000-2001 (SCHIP, 2000 controls)/5                    0.000000   
2000-2001 (1990 controls)/6                          -0.010667   
1992-1993 (1990 controls)/6                          -0.072879   
1992-1993 (1980 controls)/7                          -0.010241   
1975-1976/8                                          -0.168423   
1970-1971/8                                          -0.031542   
2013-2014 (98,000 address file)/1                    -0.999504   
2010-2011 (2010 controls)/2                           0.000000   
2010-2011 (2000 controls)/3                           0.000000   
2009-2010 (2010 controls)/2                           0.000000   
2009-2010 (2000 controls)/3                           0.000000   
2004-2005/4                                           0.000000   
2000-2001 (2000 controls)/3                           0.000000   
2000-2001 (SCHIP, 2000 controls)/5                    0.000000   
2000-2001 (1990 controls)/6                           0.000000   
1992-1993 (1990 controls)/6                           0.000000   
1992-1993 (1980 controls)/7                           0.000000   
1975-1976/8                                           0.000000   
1970-1971/8                                           0.000000   

                                    Same residence (non-movers)  Total movers  \
Mobility period                                                                 
2013-2014 (98,000 address file)/1                           NaN           NaN   
2010-2011 (2010 controls)/2                           -0.023065     -0.018021   
2010-2011 (2000 controls)/3                           -0.002511      0.001056   
2009-2010 (2010 controls)/2                           -0.014817      0.067569   
2009-2010 (2000 controls)/3                           -0.001677      0.002537   
2004-2005/4                                           -0.058175      0.062547   
2000-2001 (2000 controls)/3                           -0.043096     -0.022087   
2000-2001 (SCHIP, 2000 controls)/5                    -0.003715      0.022509   
2000-2001 (1990 controls)/6                           -0.007199     -0.031190   
1992-1993 (1990 controls)/6                           -0.103957      0.115370   
1992-1993 (1980 controls)/7                           -0.007334     -0.024386   
1975-1976/8                                           -0.177199     -0.124976   
1970-1971/8                                           -0.043649      0.024787   
2013-2014 (98,000 address file)/1                     -0.999460     -0.999695   
2010-2011 (2010 controls)/2                           -0.001130      0.008696   
2010-2011 (2000 controls)/3                            0.000000      0.000000   
2009-2010 (2010 controls)/2                           -0.010181      0.077586   
2009-2010 (2000 controls)/3                            0.000000      0.000000   
2004-2005/4                                           -0.016000      0.112000   
2000-2001 (2000 controls)/3                           -0.002933      0.018196   
2000-2001 (SCHIP, 2000 controls)/5                    -0.003715      0.022509   
2000-2001 (1990 controls)/6                            0.003506     -0.020744   
1992-1993 (1990 controls)/6                           -0.032953      0.199608   
1992-1993 (1980 controls)/7                            0.002410     -0.011765   
1975-1976/8                                           -0.010817      0.053571   


- ***what we are looking for***:
    - large % changes between same years (Mobility period(s)) due to controls
- ***notes***:
    - reset is after 'Mobility period' == 1970-1971/8
    - differences don't seem all too terrible 
- ***actions***:
    - asdf

In [26]:
# gather periods
periods = [ period[:9] for period in n_c_y_df[ 'Mobility period' ] ]
# periods = [ period for period in n_c_y_df[ 'Mobility period' ] ]

for period in periods:    
    a = n_c_y_df[n_c_y_df['Mobility period'].str.contains( period )].set_index( 'Mobility period' ).pct_change()

# compare 
def compare_mobility():
    for _ in range(len(periods)):
        a = str(n_c_y_df['Mobility period'][_:_+1])
        b = periods[_]
        q = a
        a = q[5:14]
        c = a==b
        if c:
            print( a ,' | ', b ,' | ', c )
        else:
            a = q[6:15]
            c = a==b
            if c:
                print( a ,' | ', b ,' | ', c )
            else:
                a = q[7:16]
                c = a==b
                if c:
                    print( a ,' | ', b ,' | ', c )
                else:
                    print( a ,' | ', b ,' | ', c ,' | ', q )

                
def true_true():
    q = str(n_c_y_df['Mobility period'][0:1])[5:14] == periods[0]
    w = str(n_c_y_df['Mobility period'][1:2])[5:14] == periods[1]
    return q,w
# true_true()

In [194]:
def comps(  show=True, only=False , proof=False , threshold=0.005 , count=False ):
    # set variable of index 
    iix = i_mobility_change.index
    
    # non-threshold-meeting instances
    nt=0
    # record instances that don't paired
    np=[]
    # record instances that pair
    ip=[]
    # count total 
    tc=0
    # count number of pairs
    pairs=0
    # instances of only positive threshold
    p=0
    # instances of only negative threshold
    n=0
    # instances of both positive and negative threshold
    b=0
    
    # for the length of the index -1 (last row unable to compare to next indexed row)
    for _ in range( len( iix ) - 1 ):
        # update total count
        tc+=1
        
        # year range (from 'Mobility period') for row at this index is same as row at next index
        if iix[ _ ][ :9 ] == iix[ _+1 ][ :9 ]:
            # we have a pair, so update count
            pairs+=1
            # add instance to list of instances which can pair
            ip.append( iix[ _ ] )
            # add paired instance to list of instances which can pair (not for count, do not mind duplicates)
            ip.append( iix[ _+1 ] )
            
            
            # determine how many of the changes are 0.5%+ and positive
            meets = i_mobility_change.iloc[ _+1 ] >= threshold
            # determine how many of the changes are 0.5%+ and negative
            n_meets = i_mobility_change.iloc[ _+1 ] <= -(threshold)

            # if we only want threshold instances 
            if only==True:
                
                # if at least one value meets positive threshold
                if meets.sum()>=1:
                    
                    # check for negative threshold
                    if n_meets.sum()>=1:
                        # keep count of how many times this has happened
                        b+=1
                        # output heads up on exceeding both + and - with dual thresh count
                        if show:
                            print(f'*+-*  above&below={b} *+-*')
                        
                    # no instnaces of negative threshold
                    else:
                        # update number of positive threshold only
                        p+=1
                        if show:
                            print(f'+** above only={p} **+')
                        
                    # output: 'Mobility period' of top then bottom column, change values
                    if show:
                        print( iix[ _ ] ,'\n', iix[ _+1 ] ,'\n', i_mobility_change.iloc[ _+1 ] ,'\n\n')
                
                # at least one value meets negative threshold and none of the values met the positive threshold 
                elif n_meets.sum()>=1:
                    # update number of negative threshold only
                    n+=1
                    if show:
                        print(f'-**  below only={n} **-')
                        # output: 'Mobility period' of top then bottom column, change values
                        print( iix[ _ ] ,'\n', iix[ _+1 ] ,'\n', i_mobility_change.iloc[ _+1 ] ,'\n\n')
                
                # if not meeting positive or negative threshold
                else:
                    # update count of pairs failing to meet threshold
                    nt+=1
            
            # output 'Mobility period' of top row then of bottom
            # along with the percent change from top row to bottom for each column
            else:
                if show:
                    print( iix[ _ ] ,'\n', iix[ _+1 ] ,'\n', i_mobility_change.iloc[ _+1 ] ,'\n\n')
        
        # if no pair (or; otherwise, bad measure) 
        else:
            # update list of non-pair instances
            np.append( iix[ _ ] )
            
            # if we ask for proof
            if proof==True:
                # output 'Mobility period' of row in question then for row below
                if show:
                    print( iix[ _ ], '\n' , iix[ _+1 ] ,'\n\n')
            # if we ask for simple proof
            elif proof=='simple':
                # output index and corresponding 'Mobility period'
                if show:
                    print( _ , ' | ', iix[ _ ],'\n\n' )
    
    # determine number of instances that were never paired
    np = [ i for i in np if i not in ip ]
    
    # if we request the counts
    if count:      
        # sum non-paired instances
        npi = 0
        for _p in np:
            if _p not in ip:
                npi+=1
        # sum total threshold hits 
        th = p + n + b
        # return how many times the threshold was hit
        return f'number_pairs={pairs}, met_threshold={th}, dont_meet={nt}, paired={len(ip)}, non_paired={npi}, total_count={tc}'


# let's see
comps(threshold=0.005,only=True,count=True) 

*+-*  above&below=1 *+-*
2010-2011 (2010 controls)/2 
 2010-2011 (2000 controls)/3 
 Total, 1 year old and over                 -0.002098
Same residence (non-movers)                -0.002511
Total movers                                0.001056
Different residence in the United States    0.001856
Same county                                -0.000214
Different county                            0.006401
Same state                                  0.007498
Different state                             0.004836
Movers from abroad                         -0.023985
Name: 2010-2011 (2000 controls)/3, dtype: float64 


-**  below only=1 **-
2009-2010 (2010 controls)/2 
 2009-2010 (2000 controls)/3 
 Total, 1 year old and over                 -0.001148
Same residence (non-movers)                -0.001677
Total movers                                0.002537
Different residence in the United States    0.003703
Same county                                 0.004130
Different county                      

'number_pairs=10, met_threshold=10, dont_meet=0, paired=20, non_paired=7, total_count=25'

- ***notes***:
    - 7 instances of changes greater than 0.5%
        - ***key***
            - |change| < 0.5%
                - likely not an influence control decision
                - not noted in this md
                - mentioned in conclusion of study 
            - 0.5% <= |change| < 1.0%
                - likely ok; needs to be considered moving forward, noted
            - 1.0% < |change| < 5.0% 
                - requires investigation && notation
            - |change| >= 5.0% 
                - requires thorough investigation && notation
            - 0.5%
                - all pairs meeting the 0.5% threshold also meet the 1.15% threshold
                    - 9 pairs hit 1.16% threshold
                    - 8 pairs hit 3.905% threshold 
                    - 6 pairs hit 3.959% threshold
                    - 5 pairs hit 4.0% threshold
                    - 4 pairs hit 5.0% threshold
                    - 3 paris hit 5.5% threshold
                    - 2 pairs hit 6% threshold
                    - 1 pair hits 16.666666666666664% threshold 
                        - 1992-1993 
                            - (1990 controls)/6 vs (1980 controls)/7 
                                - Movers from abroad >> -0.166667 >> 16.6667%
                - 0.5% was chosen because 1% seemed too high
                - many instances in these tables are measurements < 10,000 , some < 1,000
                    - while 1% is 1%
                        - losing 9 out of 900 or 9 out of 9,000 seems more impactful than 1,000 out of 100,000
                        - cities of 2,500,000 are (usually) not dependent on one business 
                        - areas of 250,000 may depend heavily on one business 
        - ***2 instances of only positive threshold being met***
            - 2010-2011 
                - (2010 controls)/2 vs (2000 controls)/3
                    - Different residence in the United States >> 0.008929 >> 0.8929%
                        - likely manageable 
                    - Same state >> 0.052632 >>  5.2632%
                        - **requires thorough investigation**
            - 2010-2011 
                - (2010 controls)/2 vs (2000 controls)/3
                    - Different residence in the United States >> 0.008264 >> 0.8264%
                        - likely manageable, 7.448% less than prior
                    - Same county >> 0.011628 >> 1.1628%
                        - requires investigation
        - ***3 instances of only negative threshold being met***
            - 2009-2010
                - (2010 controls)/2 vs (2000 controls)/3 
                    - Movers from abroad >> -0.039594 >> -3.9594%
                        - requires investigation
            - 1992-1993 
                - (1990 controls)/6 vs (1980 controls)/7 
                    - Total, 1 year old and over >> -0.010241 >> -1.0241%
                        - requires investigation
                    - Same residence (non-movers) >> -0.007334 >> -0.7334%
                        - likely managable
                    - Total movers >> -0.024386 >> -2.4386%
                        - requires investigation
                    - Different residence in the United States >> -0.023043 >> -2.3043%
                        - requires investigation
                    - Same county >> -0.026734 >> -2.6734%
                        - requires investigation
                    - Different county >> -0.016247 >> -1.6247%
                        - requires investigation
                    - Same state >> -0.015277 >> -1.5277%
                        - requires investigation
                    - Different state >> -0.017206 >> -1.7206%
                        - requires investigation
                    - Movers from abroad >> -0.064516 >> 6.4516%
                        - **requires thorough investigation**
            - 1992-1993 
                - (1990 controls)/6 vs (1980 controls)/7 
                    - Total movers >> -0.011765 >> -1.1765%
                        - requires investigation
                    - Different residence in the United States >> -0.012121 >> -1.2121%
                        - requires investigation
                    - Same county >> -0.018692 >> -1.8692%
                        - requires investigation
                    - Movers from abroad >> -0.166667 >> 16.6667%
                        - **requires thorough investigation**
        - ***5 instances of positive and negative thresholds being met***
            - 2010-2011 
                - (2010 controls)/2 vs (2000 controls)/3 
                    - Different county >> 0.006401 >> 0.6401%
                        - likely managable
                    - Same state >> 0.007498 >> 0.7498%
                        - likely managable
                    - Movers from abroad >> -0.023985 >> -2.3985%
                        - requires investigation
            - 2000-2001 
                - (2000 controls)/3 vs (SCHIP, 2000 controls)/5 
                    - Total movers >> 0.022509 >> 2.2509%
                        - requires investigation
                    - Different residence in the United States >> 0.023677 >> 2.3677%
                        - requires investigation
                    - Same county >> 0.039055 >> 3.9055%
                        - requires investigation
                    - Same state >> 0.030199 >> 3.0199%
                        - requires investigation
                    - Different state >> -0.025826 >> -2.5826%
                        - requires investigation
            - 2000-2001 
                - (SCHIP, 2000 controls)/5 vs (1990 controls)/6 
                    - Total, 1 year old and over >> -0.010667 >> -1.0667%
                        - requires investigation
                    - Total movers >> -0.031190 >> -3.1190%
                        - requires investigation
                    - Different residence in the United States >> -0.029895 >> -2.9895%
                        - requires investigation
                    - Same county >> -0.043515 >> -4.3515%
                        - requires investigation
                    - Same state >> -0.031756 >> -3.1756%
                        - requires investigation
                    - Different state >> 0.012793 >> 1.2793%
                        - requires investigation
                    - Movers from abroad >> -0.059361 >> -5.9361%
                        - **requires thorough investigation**
            - 2000-2001 
                - (2000 controls)/3 vs (SCHIP, 2000 controls)/5 
                    - Total movers >> 0.022509 >> 2.2509%
                        - requires investigation
                    - Different residence in the United States >> 0.022308
                        - requires investigation
                    - Same county >> 0.039055
                        - requires investigation
                    - Same state >> 0.030199
                        - requires investigation
                    - Different state >> -0.025826
                        - requires investigation
            - 2000-2001 
                - (SCHIP, 2000 controls)/5 vs (1990 controls)/6 
                    - Total movers >> -0.020744 >> -2.0744%
                        - requires investigation
                    - Different residence in the United States >> -0.018122 >> -1.8122%
                        - requires investigation
                    - Same county >> -0.033201 >> -3.3201%
                        - requires investigation
                    - Same state >> -0.021316
                        - requires investigation
                    - Different state >> 0.023714 >> 2.3714%
                        - requires investigation
                    - Movers from abroad >> -0.049219 >> -4.9219%
                        - requires investigation
- ***conclusions***:
    - unless the change is greater than 5%, we do not have time to care
- ***actions***:
    - reevaluate with threshold == 0.05

In [195]:
comps(threshold=0.05,only=True,count=True) 

-**  below only=1 **-
2000-2001 (SCHIP, 2000 controls)/5 
 2000-2001 (1990 controls)/6 
 Total, 1 year old and over                 -0.010667
Same residence (non-movers)                -0.007199
Total movers                               -0.031190
Different residence in the United States   -0.029895
Same county                                -0.043515
Different county                           -0.009766
Same state                                 -0.031756
Different state                             0.012793
Movers from abroad                         -0.059361
Name: 2000-2001 (1990 controls)/6, dtype: float64 


-**  below only=2 **-
1992-1993 (1990 controls)/6 
 1992-1993 (1980 controls)/7 
 Total, 1 year old and over                 -0.010241
Same residence (non-movers)                -0.007334
Total movers                               -0.024386
Different residence in the United States   -0.023043
Same county                                -0.026734
Different county                  

'number_pairs=10, met_threshold=4, dont_meet=6, paired=20, non_paired=7, total_count=25'

- ***notes***:
    - negative
        - 3/4 (75%) of changes >= 5% are negative only
            - 2000-2001 , /5 vs /6
                - Movers from abroad >> -0.059361 >> -5.9361%
            - 1992-1993 , /6 vs /7
                - Movers from abroad >> -0.064516 >> -6.4516%
            - 1992-1993 , /6 vs /7
                - Movers from abroad >> -0.166667 >> -16.6667%
        - all 
            - occour in Movers from abroad
            - are changed > 1% (grand total ; e.g. x=0.0075 , y=0.0064) higher than 2nd highest change
            - have these columns negative w/ |change| > 1%
                - Total movers 
                - Different residence in the United States  
                - Same county                                
    - positive
        - 1/4 (25%) of changes >= 5% are positive only
            - 2010-2011 , /2 vs /3
            - Same state >> 0.052632 >> 5.2632%
        - only ONE other affected value
            - Different residence in the United States >> 0.008929 >> 0.8929%
                - Same state = 4.3703% (grand total) difference (589.450106394893% larger)
- **knowledge**:
    - when Total movers decrease, the following may as well
        - Different residence in the United States
        - Same county 
        - Movers from abroad
- ***actions***:
    - y2k filter != SCHIP
    - determine if we want more or less Movers from abroad

4.3703